include/graybat/Cage.hpp

```diff
@@ -683,6 +692,29 @@ namespace graybat {
     template<typename T_CommunicationPolicy, typename T_GraphPolicy>
     template<typename T>
     auto Cage<T_CommunicationPolicy, T_GraphPolicy>::
+	recv (const Edge &edge, std::future<T>& data)
+	-> void
+	{
+		VAddr srcVAddr = locateVertex(edge.source);
+		std::promise<T> dataPromise;
+		data = dataPromise.get_future();
+
+		// The reason for the std::promise and std::thread over std::async ist the destructor behaviour of std::async
+		// For std::async the destructor is waiting for the completion of the asyncThread, which may lead to a permanent lock
+		// The future from std::promise does not block on the completion of the execution thread.
+		std::thread asyncRecvThread([&](){
+			T result;
+			Event e = comm->asyncRecv(srcVAddr, edge.id, graphContext, result);
+			e.wait();
+			dataPromise.set_value(result);
+		});
+		asyncRecvThread.detach();
+	}
```

Comment:
I thought a little bit more about this issue. I think your solution is up to now okay. The only thing I would change is that I would push the thread into a vector. The optimal solution would be a callback interface for the communication policies e.g.:
```
template <typename T_CommunicationPolicy, typename T_GraphPolicy, typename SerializationPolicy>
template <typename T>
auto Cage<T_CommunicationPolicy, T_GraphPolicy, SerializationPolicy>::asyncRecv(const Edge& edge, T& data)
    -> std::future<T>
{
    VAddr srcVAddr = locateVertex(edge.source);

    std::promise<void> dataReceived;

    decltype(auto) skeleton = SerializationPolicy::prepare(data);

    auto onReceived = [&data, skeleton]() {
        SerializationPolicy::restore(data, skeleton);
        dataReceived.set_value();
    };

    CPEvent e = communicator->asyncRecv(srcVAddr, edge.id, graphContext, skeleton, onReceived);

    return dataReceived.get_future();
}
```
I am not sure now how hard it is to provide this callback interface for the mpi and the zmq policies yet.

So as a workaround I would favour your solution now, but refactor all the threading into its own class (or policy). Would you like to do this refactoring?